README.md
# malabar_hijri

[![pub package](https://img.shields.io/pub/v/malabar_hijri_date.svg)](https://pub.dev/packages/malabar_hijri_date)
<!-- [![Build Status](https://travis-ci.org/dart-lang/malabar_hijri_date.svg?branch=master)](https://travis-ci.org/dart-lang/malabar_hijri_date) -->

A Flutter Package for Hijri Date in Malabar of Kearala India, restructured every night of hilal sight

## Usage

#### Convert Gregorian date to Malabar Hijri:

 ```dart
  int year= DateTime.now().year;
  int month= DateTime.now().month;
  int day= DateTime.now().day;
  gregorianToHijri(year, month, day);
  /// result will be like: 2020, 12, 28
```


CHANGELOG.md
## [0.0.1] - TODO: Add release date.

* Initiate date api.


lib/src/array.dart
import 'package:dio/dio.dart';

final Map<int, String> monthNames = {
  1: 'Muharram',
  2: 'Safar',
  3: 'Rabi\' Al-Awwal',
  4: 'Rabi\' Al-Thani',
  5: 'Jumada Al-Awwal',
  6: 'Jumada Al-Thani',
  7: 'Rajab',
  8: 'Sha\'aban',
  9: 'Ramadan',
  10: 'Shawwal',
  11: 'Dhu Al-Qi\'dah',
  12: 'Dhu Al-Hijjah'
};

final Map<int, String> monthShortNames = {
  1: 'Muh',
  2: 'Saf',
  3: 'Rab1',
  4: 'Rab2',
  5: 'Jum1',
  6: 'Jum2',
  7: 'Raj',
  8: 'Sha\'',
  9: 'Ram',
  10: 'Shaw',
  11: 'DhuQ',
  12: 'DhuH'
};

final Map<int, String> wdNames = {
  7: "Sunday",
  1: "Monday",
  2: "Tuesday",
  3: "Wednesday",
  4: "Thursday",
  5: "Friday",
  6: "Saturday"
};

final Map<int, String> shortWdNames = {
  7: "Sun",
  1: "Mon",
  2: "Tue",
  3: "Wed",
  4: "Thu",
  5: "Fri",
  6: "Sat"
};

Future<dynamic> res() async {
  var data = await Dio()
      .get('https://malabar-ijri-api.herokuapp.com/array')
      .then((value) => value.data[0]['data']);
  return data;
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:malabar_hijri_date/malabar_hijri_date.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MalabarHijriHome(),
    );
  }
}

class MalabarHijriHome extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: FutureBuilder(
          future: gregorianToHijri(2020, 12, 19),
          builder: (context, snapshot) {
            if (!snapshot.hasData)
              return Center(child: CircularProgressIndicator());
            return Center(
              child: Text(snapshot.data),
            );
          }),
    );
  }
}


lib/malabar_hijri.dart
library malabar_hijri;

import 'package:malabar_hijri/src/array.dart';

export 'package:malabar_hijri/src/malabar_hirjri_date.dart';

int lengthOfMonth;
int hDay;
int hMonth;
int hYear;
int wkDay;
String longMonthName;
String shortMonthName;
String dayWeName;
Map<int, int> adjustments;

_gMod(int n, int m) {
  return ((n % m) + m) % m;
}

gregorianToHijri(pYear, pMonth, pDay) async{
  return await res().then((ummAlquraDateArray){
    var day = (pDay);
    var month = (pMonth);
    var year = (pYear);
    var m = month;
    var y = year;
    if (m < 3) {
      y -= 1;
      m += 12;
    }

    var a = (y / 100).floor();
    var jgc = a - (a / 4.0).floor() - 2;
    var cjdn = (365.25 * (y + 4716)).floor() +
        (30.6001 * (m + 1)).floor() +
        day -
        jgc -
        1524;
    a = ((cjdn - 1867216.25) / 36524.25).floor();
    jgc = a - (a / 4.0).floor() + 1;
    var b = cjdn + jgc + 1524;
    var c = ((b - 122.1) / 365.25).floor();
    var d = (365.25 * c).floor();
    month = ((b - d) / 30.6001).floor();
    day = (b - d) - (30.6001 * month).floor();

    if (month > 13) {
      c += 1;
      month -= 12;
    }
    month -= 1;
    year = c - 4716;
    var mcjdn = cjdn - 2400000;
    var i;
    for (i = 0; i < ummAlquraDateArray.length; i++) {
      if (ummAlquraDateArray[i] > mcjdn) break;
    }

    var iln = i + 16260;
    var ii = ((iln - 1) / 12).floor();
    var iy = ii + 1;
    var im = iln - 12 * ii;
    var id = mcjdn - ummAlquraDateArray[i - 1] + 1;
    var ml = ummAlquraDateArray[i] - ummAlquraDateArray[i - 1];
    print(ummAlquraDateArray[i]);
    var lengthOfMonth = ml;
    var wd = _gMod(cjdn + 1, 7);
    wkDay = wd == 0 ? 7 : wd;
    return '$iy, $im, $id';
  });

}


lib/src/malabar_hirjri_date.dart
import 'package:flutter/material.dart';

import '../malabar_hijri.dart';
import 'array.dart';

class MalabarHijriDate extends StatefulWidget {
  final String text;

  const MalabarHijriDate({Key key, this.text}) : super(key: key);

  @override
  _MalabarHijriDateState createState() => _MalabarHijriDateState();
}

class _MalabarHijriDateState extends State<MalabarHijriDate> {


  @override
  Widget build(BuildContext context) {

    ii()async{
      print(await gregorianToHijri(2020, 12, 18));
    }
    ii();
    return Semantics(
      container: false,
      button: false,
      enabled: false,
      child: new ConstrainedBox(
        constraints: BoxConstraints(minWidth: 200),
        child: new Material(
          child: Container(color:Colors.green,child: Text('ljl')),
        ),
      ),
    );
  }
}


lib/malabar_hijri_date.dart
library malabar_hijri_date;

import 'package:malabar_hijri_date/src/array.dart';

int lengthOfMonth;
int hDay;
int hMonth;
int hYear;
int wkDay;
String longMonthName;
String shortMonthName;
String dayWeName;
Map<int, int> adjustments;

/// generalized module function (n mod m)
_gMod(int n, int m) {
  return ((n % m) + m) % m;
}

/// convert hijri to gregorian date
gregorianToHijri(pYear, pMonth, pDay) async {
  return await res().then((ummAlquraDateArray) {
    var day = (pDay);
    var month = (pMonth);
    var year = (pYear);
    var m = month;
    var y = year;
    if (m < 3) {
      y -= 1;
      m += 12;
    }

    var a = (y / 100).floor();
    var jgc = a - (a / 4.0).floor() - 2;
    var cjdn = (365.25 * (y + 4716)).floor() +
        (30.6001 * (m + 1)).floor() +
        day -
        jgc -
        1524;
    a = ((cjdn - 1867216.25) / 36524.25).floor();
    jgc = a - (a / 4.0).floor() + 1;
    var b = cjdn + jgc + 1524;
    var c = ((b - 122.1) / 365.25).floor();
    var d = (365.25 * c).floor();
    month = ((b - d) / 30.6001).floor();
    day = (b - d) - (30.6001 * month).floor();

    /// append January and February to the previous year (i.e. regard March as
    /// the first month of the year in order to simplify leapday corrections)
    if (month > 13) {
      c += 1;
      month -= 12;
    }
    month -= 1;
    year = c - 4716;
    var mcjdn = cjdn - 2400000;
    var i;
    for (i = 0; i < ummAlquraDateArray.length; i++) {
      if (ummAlquraDateArray[i] > mcjdn) break;
    }

    var iln = i + 16260;
    var ii = ((iln - 1) / 12).floor();
    var iy = ii + 1;
    var im = iln - 12 * ii;
    var id = mcjdn - ummAlquraDateArray[i - 1] + 1;
    // var ml = ummAlquraDateArray[i] - ummAlquraDateArray[i - 1];
    print(ummAlquraDateArray[i]);
    // var lengthOfMonth = ml;
    var wd = _gMod(cjdn + 1, 7);
    wkDay = wd == 0 ? 7 : wd;
    return '$iy, $im, $id';
  });
}
